### PR TITLE
Deserialization updates

### PR DESCRIPTION
**Pull request by Shane32**
Also fixes bug with deserializing dates.

### CODE REVIEW

File: src/Transports.Subscriptions.Abstractions/Transports.Subscriptions.Abstractions.csproj [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (sungam3r) [inline src/Transports.Subscriptions.Abstractions/Transports.Subscriptions.Abstractions.csproj:2]: Encoding?

File: src/Transports.AspNetCore.SystemTextJson/InternalGraphQLRequest.cs [hunk 2]
```diff
@@ -19,13 +18,13 @@ internal sealed class InternalGraphQLRequest
...
-        public Dictionary<string, object> Variables { get; set; }
+        public Inputs Variables { get; set; }
 
         /// <remarks>
         /// Population of this property during deserialization from JSON requires
         /// <see cref="GraphQL.SystemTextJson.ObjectDictionaryConverter"/>.
         /// </remarks>
         [JsonPropertyName(GraphQLRequest.EXTENSIONS_KEY)]
-        public Dictionary<string, object> Extensions { get; set; }
+        public Inputs Extensions { get; set; }
```
Comment (sungam3r) [inline src/Transports.AspNetCore.SystemTextJson/InternalGraphQLRequest.cs:28]: 😄  I expect you do exactly that change. But what about nested dictionaries (not inputs) inside that top level property? Will they be properly parsed?

File: src/Transports.AspNetCore.NewtonsoftJson/GraphQLRequestDeserializer.cs [hunk 3]
```diff
@@ -18,7 +18,15 @@ public class GraphQLRequestDeserializer : IGraphQLRequestDeserializer
 
         public GraphQLRequestDeserializer(Action<JsonSerializerSettings> configure)
         {
-            var settings = new JsonSerializerSettings();
+            var settings = new JsonSerializerSettings()
+            {
+                DateFormatHandling = DateFormatHandling.IsoDateFormat,
+                DateParseHandling = DateParseHandling.None,
```
Comment (sungam3r) [inline src/Transports.AspNetCore.NewtonsoftJson/GraphQLRequestDeserializer.cs:24]: 👍 ping @slavautesinov

Comment (SlavaUtesinov) [reply src/Transports.AspNetCore.NewtonsoftJson/GraphQLRequestDeserializer.cs:24]: Yeah, it's another mentioned possible solution

File: src/Transports.AspNetCore.SystemTextJson/InternalGraphQLRequest.cs (hunk 2 above)
Comment (Shane32) [reply src/Transports.AspNetCore.SystemTextJson/InternalGraphQLRequest.cs:28]: Yes, but I should add a test to be sure.

File: src/Transports.AspNetCore.SystemTextJson/GraphQLRequestDeserializer.cs [hunk 4]
```diff
@@ -24,7 +24,7 @@ public class GraphQLRequestDeserializer : IGraphQLRequestDeserializer
         public GraphQLRequestDeserializer(Action<JsonSerializerOptions> configure)
         {
             // Add converter that deserializes Variables property
```
Comment (sungam3r) [inline src/Transports.AspNetCore.SystemTextJson/GraphQLRequestDeserializer.cs:26]:
```suggestion
            // Add converter that deserializes Variables and Extensions properties
```

File: src/Transports.AspNetCore.SystemTextJson/GraphQLRequestDeserializer.cs [hunk 5]
```diff
@@ -117,8 +117,8 @@ private static GraphQLRequest ToGraphQLRequest(InternalGraphQLRequest internalGr
             {
                 OperationName = internalGraphQLRequest.OperationName,
                 Query = internalGraphQLRequest.Query,
-                Inputs = internalGraphQLRequest.Variables?.ToInputs(), // must return null if not provided, not an empty dictionary
-                Extensions = internalGraphQLRequest.Extensions?.ToInputs(), // must return null if not provided, not an empty dictionary
+                Inputs = internalGraphQLRequest.Variables, // must return null if not provided, not an empty dictionary
+                Extensions = internalGraphQLRequest.Extensions, // must return null if not provided, not an empty dictionary
```
Comment (sungam3r) [inline src/Transports.AspNetCore.SystemTextJson/GraphQLRequestDeserializer.cs:121]:
```suggestion
                Inputs = internalGraphQLRequest.Variables, // must return null if not provided, not an empty Inputs
                Extensions = internalGraphQLRequest.Extensions, // must return null if not provided, not an empty Inputs
```

File: src/Transports.AspNetCore.NewtonsoftJson/GraphQLRequestDeserializer.cs [hunk 6]
```diff
@@ -18,7 +18,15 @@ public class GraphQLRequestDeserializer : IGraphQLRequestDeserializer
 
         public GraphQLRequestDeserializer(Action<JsonSerializerSettings> configure)
         {
-            var settings = new JsonSerializerSettings();
+            var settings = new JsonSerializerSettings()
```
Comment (sungam3r) [inline src/Transports.AspNetCore.NewtonsoftJson/GraphQLRequestDeserializer.cs:21]:
```suggestion
            var settings = new JsonSerializerSettings
```

File: src/Transports.AspNetCore.NewtonsoftJson/GraphQLRequestDeserializer.cs [hunk 7]
```diff
@@ -67,8 +75,8 @@ private static GraphQLRequest ToGraphQLRequest(InternalGraphQLRequest internalGr
             {
                 OperationName = internalGraphQLRequest.OperationName,
                 Query = internalGraphQLRequest.Query,
-                Inputs = internalGraphQLRequest.Variables?.ToInputs(), // must return null if not provided, not an empty dictionary
-                Extensions = internalGraphQLRequest.Extensions?.ToInputs(), // must return null if not provided, not an empty dictionary
+                Inputs = internalGraphQLRequest.Variables, // must return null if not provided, not an empty dictionary
+                Extensions = internalGraphQLRequest.Extensions, // must return null if not provided, not an empty dictionary
```
Comment (sungam3r) [inline src/Transports.AspNetCore.NewtonsoftJson/GraphQLRequestDeserializer.cs:79]:
```suggestion
                Inputs = internalGraphQLRequest.Variables, // must return null if not provided, not an empty Inputs
                Extensions = internalGraphQLRequest.Extensions, // must return null if not provided, not an empty Inputs
```

File: GraphQL.Server.sln [hunk 8]
```diff
@@ -78,6 +78,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "ApiApprovalTests", "tests\A
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "All", "src\All\All.csproj", "{B39BC064-B96A-442D-89A0-79B750F88ADD}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Transports.AspNetCore.Tests", "tests\Transports.AspNetCore.Tests\Transports.AspNetCore.Tests.csproj", "{1857A25B-C4A3-4332-9077-3F2516AC6059}"
```
Comment (sungam3r) [inline GraphQL.Server.sln:81]: Add it to `Upload coverage to codecov` step files.

File: tests/Transports.AspNetCore.Tests/NewtonsoftJsonTests.cs [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+        [Fact]
+        public async Task Decodes_Multiple_Queries()
+        {
+            var ret = await Deserialize(@"[{""query"":""abc""},{""query"":""def""}]");
+            ret.Batch.Length.ShouldBe(2);
+            ret.Batch[0].Query.ShouldBe("abc");
+            ret.Batch[1].Query.ShouldBe("def");
+        }
+        private async Task<GraphQLRequestDeserializationResult> Deserialize(string jsonText)
```
Comment (sungam3r) [inline tests/Transports.AspNetCore.Tests/NewtonsoftJsonTests.cs:75]:
```suggestion
        }

        private async Task<GraphQLRequestDeserializationResult> Deserialize(string jsonText)
```

File: tests/Transports.AspNetCore.Tests/NewtonsoftJsonTests.cs [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+using System.Threading.Tasks;
+using Shouldly;
+using Xunit;
+
+namespace GraphQL.Server.Transports.AspNetCore.Tests
+{
+    public partial class NewtonsoftJsonTests
+    {
+        [Fact]
+        public async Task Decodes_Request()
```
Comment (sungam3r) [inline tests/Transports.AspNetCore.Tests/NewtonsoftJsonTests.cs:12]: Need a test with complex variables/extensions.

File: tests/Transports.AspNetCore.Tests/SystemTextJsonTests.cs [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+        [Fact]
+        public async Task Decodes_Multiple_Queries()
+        {
+            var ret = await Deserialize(@"[{""query"":""abc""},{""query"":""def""}]");
+            ret.Batch.Length.ShouldBe(2);
+            ret.Batch[0].Query.ShouldBe("abc");
+            ret.Batch[1].Query.ShouldBe("def");
+        }
+        private async Task<GraphQLRequestDeserializationResult> Deserialize(string jsonText)
```
Comment (sungam3r) [inline tests/Transports.AspNetCore.Tests/SystemTextJsonTests.cs:75]:
```suggestion
        }

        private async Task<GraphQLRequestDeserializationResult> Deserialize(string jsonText)
```

File: tests/Transports.AspNetCore.Tests/SystemTextJsonTests.cs [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+        [Fact]
+        public async Task Extensions_Null_When_Not_Provided()
+        {
+            var ret = await Deserialize(@"{""variables"":{""date"":""2015-12-22T10:10:10+03:00""}}");
+            ret.Single.Extensions.ShouldBeNull();
+        }
+
+        [Fact]
+        public async Task Name_Matching_Is_Case_Sensitive()
```
Comment (sungam3r) [inline tests/Transports.AspNetCore.Tests/SystemTextJsonTests.cs:61]: Differs from Newtonsoft default behavior. Would you write a documentation PR ?

**Review by sungam3r — CHANGES_REQUESTED** [review]
Done. Minor issues detected.

File: tests/Transports.AspNetCore.Tests/SystemTextJsonTests.cs (hunk 12 above)
Comment (Shane32) [reply tests/Transports.AspNetCore.Tests/SystemTextJsonTests.cs:61]: I would like to, if I have time.

Comment (Shane32) [reply tests/Transports.AspNetCore.Tests/SystemTextJsonTests.cs:61]: In the main repo that is

Comment (sungam3r) [reply tests/Transports.AspNetCore.Tests/SystemTextJsonTests.cs:61]: https://github.com/graphql-dotnet/graphql-dotnet/issues/2529

File: src/Transports.AspNetCore.SystemTextJson/InternalGraphQLRequest.cs (hunk 2 above)
Comment (Shane32) [reply src/Transports.AspNetCore.SystemTextJson/InternalGraphQLRequest.cs:28]: Done

File: src/Transports.Subscriptions.Abstractions/Transports.Subscriptions.Abstractions.csproj (hunk 1 above)
Comment (Shane32) [reply src/Transports.Subscriptions.Abstractions/Transports.Subscriptions.Abstractions.csproj:2]: 🤷‍♂️ I didn't purposefully make a change, or know how to change the encoding, or know what it was changed to, or what it was.

File: tests/Transports.AspNetCore.Tests/NewtonsoftJsonTests.cs (hunk 10 above)
Comment (Shane32) [reply tests/Transports.AspNetCore.Tests/NewtonsoftJsonTests.cs:13]: Done for both nested dictionary and array

File: GraphQL.Server.sln (hunk 8 above)
Comment (Shane32) [reply GraphQL.Server.sln:81]: I made a change; you'll need to verify that the change is correct.

File: .github/workflows/test.yml [hunk 13]
```diff
@@ -63,7 +63,7 @@ jobs:
         if: ${{ startsWith(matrix.os, 'ubuntu') }}
         uses: codecov/codecov-action@v1.2.1
         with:
-          files: .coverage/GraphQL.Server.Authorization.AspNetCore.Tests/coverage.net5.opencover.xml,.coverage/GraphQL.Server.Samples.Server.Tests/coverage.netcoreapp3.1.opencover.xml,.coverage/GraphQL.Server.Transports.Subscriptions.Abstractions.Tests/coverage.net5.opencover.xml,.coverage/GraphQL.Server.Transports.Subscriptions.WebSockets.Tests/coverage.net5.opencover.xml
+          files: .coverage/GraphQL.Server.Authorization.AspNetCore.Tests/coverage.net5.opencover.xml,.coverage/GraphQL.Server.Samples.Server.Tests/coverage.netcoreapp3.1.opencover.xml,.coverage/GraphQL.Server.Transports.Subscriptions.Abstractions.Tests/coverage.net5.opencover.xml,.coverage/GraphQL.Server.Transports.Subscriptions.WebSockets.Tests/coverage.net5.opencover.xml,.coverage/GraphQL.Server.Transports.AspNetCore.Tests/coverage.net5.opencover.xml
```
Comment (Shane32) [inline .github/workflows/test.yml:66]: Please review this change.

File: GraphQL.Server.sln (hunk 8 above)
Comment (sungam3r) [reply GraphQL.Server.sln:81]: CodeCov bot update this comment https://github.com/graphql-dotnet/server/pull/569#issuecomment-826001926 , good

File: .github/workflows/test.yml (hunk 13 above)
Comment (sungam3r) [reply .github/workflows/test.yml:66]: 👌 
изображение

**Review by sungam3r — APPROVED** [review]
Bump to 5.0.3

**Comment by Shane32** [comment]
Merge now or wait until next published version of GraphQL.NET so that the dependencies are correct?

**Comment by sungam3r** [comment]
Merge only with stable dependencies.

**Comment by sungam3r** [comment]
I do not remember anything about it. All files are marked as viewed, so feel free to merge.

**Comment by Shane32** [comment]
Merging now that GraphQL.NET 4.6.0 was released (no changes).  Don't know why I didn't merge sooner; must have forgot about it.

**Pull request merged**